src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -2280,6 +2280,23 @@ private static void RegisterSerializationProviders(List<TypeInfo> providerTypes)
                         logger.Error(ErrorCode.SerMgr_ErrorLoadingAssemblyTypes, "Failed to create instance of type: " + type.FullName, exception);
                     }
                 });
+
+            // for the outside runtime client, a type may not be registered in the serialization manager before attempting to deserialize
+            // which ends up causing deserialization to fail.  to fix this, we need to register the type if it's supported by any serializer.
+            Action<Assembly> registerAction = 
+                assembly =>
+                {
+                    foreach (var type in assembly.GetTypes().Where(
+                        type => 
+                        (type.IsClass || type.IsValueType) && !type.IsAbstract &&
+                        externalSerializers.Any(ser => ser.IsSupportedType(type))))
+                    {
+                        SerializationManager.Register(type);
+                    }
+                };
+
+            AppDomain.CurrentDomain.AssemblyLoad += (o, e) => registerAction(e.LoadedAssembly);
+            Array.ForEach(AppDomain.CurrentDomain.GetAssemblies(), registerAction);
```

Comment:
The only way of doing this is to add some methods to the TypeUtilities class that allows you to resolve the type from the OrleansTypeKeyString.  Currently, it's just a dictionary lookup which is why the types need to be registered prior to deserialization.